File: GoldenHammer/GoldenHammer.js
// The devoloper uses Array.protoype.forEach() method to solve every problem
/* const people = [
	{ name: 'Alice', age: 30, job: 'Developer' },
	{ name: 'Bob', age: 25, job: 'Designer' },
	{ name: 'Charlie', age: 35, job: 'Manager' }
];
// using froEach for everthing, even simple operations
function getSumAges(data) {
	let sum = 0;
	data.forEach((person) => (sum += person.age));
	return sum;
}
function findOldestPerson(data) {
	let oldest = data[0];
	data.forEach((person) => {
		if (person.age > oldest.age) oldest = person;
	});
	return oldest;
}

function filterByJob(data, job) {
	const filtered = [];
	data.forEach((person) => {
		if (person.job === job) {
			filtered.push(person);
		}
	});
	return filtered;
}

console.log(getSumAges(people)); // 90
console.log(findOldestPerson(people)); // { name: 'Charlie', age: 35, job: 'Manager' }
console.log(filterByJob(people, 'Developer')); // { name: 'Alice', age: 30, job: 'Developer' },
 */

// Above code overusing Array.prototype.forEach()
const people = [
	{ name: 'Alice', age: 30, job: 'Developer' },
	{ name: 'Bob', age: 25, job: 'Designer' },
	{ name: 'Charlie', age: 35, job: 'Manager' }
];

function getSumAges(data) {
	// Usign built-in Javascript Array methods when apprpriate
	return data.reduce((sum, person) => sum + person.age, 0);
}
function findOldestPerson(data) {
	// Usign built-in Javascript Array methods when apprpriate
	return data.reduce((oldest, person) => (person.age > oldest.age ? person : oldest));
}

function filterByJob(data, job) {
	return data.filter((person) => person.job === job);
}

console.log(getSumAges(people)); // 90
console.log(findOldestPerson(people)); // { name: 'Charlie', age: 35, job: 'Manager' }
console.log(filterByJob(people, 'Developer')); // { name: 'Alice', age: 30, job: 'Developer' },

File: GoldenHammer/GoldenHammer.py
# The developer oversues the for loop to solve every problem in the script

# people =[
#     {'name': 'Alice' , 'age': 30, 'job' : 'Developer'},
#     {'name': 'Jhon' , 'age': 25, 'job' : 'Designer'},
#     {'name': 'Charlie' , 'age': 35, 'job' : 'Manager'}
# ]

# def get_sum_of_ages(data):
#     sum_of_ages = 0
#     for person in data:
#         sum_of_ages += person['age']
#     return sum_of_ages

# def find_oldest_person(data):
#     oldest_person = data[0]
#     for person in data:
#         if person['age'] > oldest_person['age']:
#             oldest_person = person
#     return oldest_person

# def filter_by_jobs(data, job):
#     filtered = []
#     for person in data:
#         if person['job'] == job:
#             filtered.append(person)
#     return filtered

# print(get_sum_of_ages(people))
# print(find_oldest_person(people))
# print(filter_by_jobs(people, 'Developer'))

# Now we are refactoring this code by avoiding to use 'for' loop and will be using built-in python functions because they are more appropriate and easier to understand 

people =[
    {'name': 'Alice' , 'age': 30, 'job' : 'Developer'},
    {'name': 'Jhon' , 'age': 25, 'job' : 'Designer'},
    {'name': 'Charlie' , 'age': 35, 'job' : 'Manager'}
]

def get_sum_of_ages(data):
    return sum((person['age'] for person in data))

def find_oldest_person(data):
    return max(data, key=lambda person: person['age'])

def filter_by_jobs(data, job):
    return [person for person in data if person['job'] == job]

print(get_sum_of_ages(people))
print(find_oldest_person(people))
print(filter_by_jobs(people, 'Developer'))